### PR TITLE
NLRFIM-169 added support to AppSwitch

### PR DESCRIPTION
**Pull request by achabala**

### CODE REVIEW

File: src/dk.nita.saml20/dk.nita.saml20/Utils/AppSwitchSerialization.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public static class AppSwitchSerialization
+    {
+        /// <summary>
+        /// Serializes as XmlElement.
+        /// </summary>
+        /// <param name="appSwitch">AppSwitch instance.</param>
+        /// <param name="owner">Parent xml document.</param>
+        /// <returns>Serialized Xml document.</returns>
+        public static XmlElement ToXmlElement(this AppSwitch appSwitch, XmlDocument owner)
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.saml20/Utils/AppSwitchSerialization.cs:21]: Why is it that the existing Serialization-class can't be used?

File: src/dk.nita.saml20/dk.nita.saml20/public-extensions-schema.xsd [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.saml20/public-extensions-schema.xsd:1]: Place it in the "Schema"-folder.

File: src/dk.nita.saml20/dk.nita.test.saml20/Saml20/Protocol/HttpRedirectBindingTest.cs [hunk 3]
```diff
@@ -104,7 +107,7 @@ public void TestRSASigning()
         public void TestParsing_01()
         {
             HttpRedirectBindingBuilder bindingBuilder = new HttpRedirectBindingBuilder();
-            string request = string.Empty.PadLeft(350, 'A')+"������";
+            string request = string.Empty.PadLeft(350, 'A')+"������";
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.test.saml20/Saml20/Protocol/HttpRedirectBindingTest.cs:110]: Are you sure about this changes?

File: src/dk.nita.saml20/dk.nita.test.saml20/app.config [hunk 4]
```diff
@@ -45,6 +45,8 @@
         <md:TelephoneNumber>70225885</md:TelephoneNumber>
       </md:ContactPerson>
     </ServiceProvider>
+    <AppSwitchReturnURL platform="Android">sp0.test-nemlog-in.dk</AppSwitchReturnURL>
+    <AppSwitchReturnURL platform = "iOS">sp1.test-nemlog-in.dk</AppSwitchReturnURL>
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.test.saml20/app.config:49]:
```suggestion
    <AppSwitchReturnURL platform="iOS">sp1.test-nemlog-in.dk</AppSwitchReturnURL>

```

File: src/dk.nita.saml20/dk.nita.saml20/Config/AppSwitchReturnURL.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    /// <summary>
+    /// Configuration element for app switch return url.
+    /// </summary>
+    public class AppSwitchReturnURL 
+    {
+        /// <summary>
+        /// AppSwitch platform. Can be either Android or iOS.
+        /// </summary>
+        [XmlAttribute(AttributeName = "platform")]
+        public string Platform;
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.saml20/Config/AppSwitchReturnURL.cs:14]: Why isn't this an enum value? Would make logic in the FindAppSwitchReturnUrlForPlatform-method simpler and the value is already suppose to be a known set of values.

I would discourage the use of string comparison in this situation.

File: src/dk.nita.saml20/dk.nita.saml20/Config/SAML20FederationConfig.cs [hunk 6]
```diff
@@ -144,7 +145,27 @@ public IDPEndPoint FindEndPoint(string endPointId)
...
+        /// <returns>String, containing the platform-specific Return URL.</returns>
+        public string FindAppSwitchReturnUrlForPlatform(string appSwitchPlatform)
+        {
+            if (string.IsNullOrWhiteSpace(appSwitchPlatform))
+            {
+                throw new ArgumentNullException(nameof(appSwitchPlatform));
+            }
 
+            return AppSwitchReturnURL.SingleOrDefault(x =>
+                x.Platform.Trim().ToLowerInvariant() == appSwitchPlatform.Trim().ToLowerInvariant())?.Value;
```
Comment (mdbech) [inline src/dk.nita.saml20/dk.nita.saml20/Config/SAML20FederationConfig.cs:167]: Suddently comparison is case insensitive.

File: Net SAML2 Service Provider Framework.docx
Comment (mdbech) [inline Net SAML2 Service Provider Framework.docx:1]: MitID not MyID.

Comment (mdbech) [inline Net SAML2 Service Provider Framework.docx:1]: I'm missing documentation of the option of using the new HTTP request parameter appSwitchPlatform for the SignonHandler.

**Review by mdbech — CHANGES_REQUESTED** [review]

File: Net SAML2 Service Provider Framework.docx
Comment (mdbech) [inline Net SAML2 Service Provider Framework.docx:1]: Please re-upload, the file is damaged and can't be opened.

Peter had similar problem with the Java version, please check with him if you have problems.

**Review by mdbech — CHANGES_REQUESTED** [review]

File: Net SAML2 Service Provider Framework.docx
Comment (achabala) [reply Net SAML2 Service Provider Framework.docx:1]: done

**Review by mdbech — APPROVED** [review]

**Pull request merged**